---
title: "R introduction: part I"
author: "Staton Lab"
date: "5/29/2017"
output: html_document
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

# Content

* [Set working directory](#set-working-directory)
* [Install packages](#install-packages)
* [Get some help](#get-some-help)
* [Data and data structures](#data-and-data-structures)
* [Subsetting operators](#subsetting-operators)
* [Factors](#factors)
* [Write data into files](#write-data-into-files)


# Set working directory
R has a working directory, which is like being in a certain location in the terminal. That location is where R will automatically look for files and write new files. You can find out where this location is with setwd.
```{r eval=FALSE}
getwd()
```

You can also set to a new location, anywhere in your file system. We want this to be a folder on your Desktop for now. Create the folder "rnaseq-workshop-2017" then run this command:
```{r eval=FALSE}
# For Mac user
setwd('~/Desktop/rnaseq-workshop-2017')
# For Windows user (Thanks to Fang!)
setwd('C:/Users/fliu21/Desktop/rnaseq-workshop-2017')
```


# Install packages
R has many packages of code that you can use. You can explore these through [CRAN](https://cran.r-project.org/). We will download some basic packages to demonstrate how to access those modules. ggplot2 is an R package that does plotting, i.e. creates very nice graphics.

### Install package from CRAN
The install.packages command will download the files through the internet and install them on your local computer (so you only need to do this once for a package).

The require command loads and attaches the package to this session. Basically, it tells R that you want to use the package in this session and to make all its functions available.

```{r eval=FALSE}
install.packages('ggplot2')
library('ggplot2')
require('ggplot2')
```

### Install bioconductor packages
Bioconductor is a collection of tools for the analysis and comprehension of high-throughput genomic data. It includes many differential gene expression statistical packages, including DESeq2. It has its own system for package installation.
    + google 'deseq2 bioconductor'
    + https://www.bioconductor.org/packages/release/bioc/html/DESeq2.html

```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
require('DESeq2')
packageVersion('DESeq2')
```

### Install package from github
This is a much less often used form of getting bioconductor packages installed into R but it can come in very handy. For example, if you need to get an older version of the software (sometimes needed for DESeq2 depending on your OS).
    + go to https://www.bioconductor.org/packages/release/bioc/html/DESeq2.html

```{r eval=FALSE}
remove.packages('DESeq2')
# From R Studio menu Session, click Restart R
require('DESeq2')

install.packages('devtools')
require('devtools')
install_github('Bioconductor-mirror/DESeq2')

## to install a specific version.
install_github('Bioconductor-mirror/DESeq2', ref = 'release-3.3')
packageVersion('DESeq2')

## read this page if this doesn't work for you:
## https://stackoverflow.com/questions/29992066/rcpp-warning-directory-not-found-for-option-l-usr-local-cellar-gfortran-4-8/29993906#29993906
```

<hr>

### Exercise

* Install the bioconductor package *Rsubread* version *release-3.4* from github.

<hr>


> ## more package installation methods
>* Install package from local source file
>    + google 'ggplot2 tar.gz'
>    + https://cran.r-project.org/web/packages/ggplot2/index.html
>```{r eval=FALSE}
>remove.packages('ggplot2')
>## restart R
>library(ggplot2)
>require('ggplot2')
>install.packages('ggplot2_2.2.1.tar.gz', repos = NULL)
>require('ggplot2')
>```
>
>* Install package from command line
>    + create an R script named `install-r-packages.R` and put the following >content into this file.
>    + Open your terminal and run command: `RScript install-r-packages.R`
>
>```{r eval=FALSE}
>pkgs = c('plyr', 'dplyr')
>install.packages(pkgs, repos = "http://cran.us.r-project.org")
>```


# How to get help

* The `?` function.
    + `?install.packages`
    + **?\`?\`**, **?\`+\`**
    + `` ?`?` ``
    + `` ?`+` ``
* The `help()` function: `help(package='DESeq2')`
* The `swirl()` package is an interactive tutorial for learning R 
* **google**

### Comments
Note that like many programming language, lines that start with a `#` are comments and are not interpreted. 
```{r eval=FALSE}
# You can write notes and annotate your code in comments!
```


# Data and data structures
## Built-in datasets

```{r}
data(package='base')
data(package='ggplot2')
data('mpg', package='ggplot2')
mpg
```

## loading data

Two most common data loading functions: 

* `read.csv()` 
* `read.table()`
    
Both `read.csv()` and `read.table()` reads a flat file (e.g., csv file or delimited file) in table format. *By default*, `read.csv()` reads comma separated values and `read.table()` reads *space or tab* separated values. But you can change the default and use these two functions interchangeably.

```{r}
?read.csv
read.csv('PBSC-comma.txt') # the file does not have to be a csv file.
read.csv('PBSC-tab.txt') # get wierd result
read.csv('PBSC-tab.txt', sep = '\t') # change the default delimiter.
read.csv('PBSC-comma.txt', header=FALSE)
```


```{r}
read.table('PBSC-tab.txt', header=TRUE) # does the header argument work?
# open the file and see what the first line look like.
read.table('PBSC-tab.txt', header=TRUE, comment.char = '') # change the default comment character

# an example use case: load gtf file
read.table('Arabidopsis_thaliana.h20.gtf', header=FALSE, sep='\t', 
         comment.char = '')

read.table('Arabidopsis_thaliana.h20.gtf', header=FALSE, sep='\t', 
         comment.char = '#')
```


<hr>

### Exercise

* Put one of your own csv, txt or gtf files in the working directory and try to read the data into R with these two functions.

<hr>

## variables

* Restrictions on variable names.
    + Technically almost no restrictions on variable names
    + But, for **Most R programmers**, variable names... 
        * start with lower case letters 
        * separate words in names with underscores
        * only use lower case letters, numbers and underscores
        
```{r}
x = 1:10
x
my_data = 1:10
my_data
my_data_2 = 1:5
my_data_2

assign(x = 'my_data', value = 3:8)
my_data

# some variable names that you should probably never use.
assign(x = '2017', value = 1:10)
'2017'
get('2017')

assign(x = '#', value = 'this is my value')
get('#')
```
    
* Assign the imported data to a variable.
```{r}
pbsc = read.csv('PBSC-comma.txt', header = TRUE)
pbsc

pbsc <- read.csv('PBSC-comma.txt', header = TRUE)

pbsc
```

You might see the `<-` assignment in people's code.  It is the same as `=`. But I prefer using the `=` operator.

## data structures

* Data frame
    + the most common data structure in R
    + has two dimensions
    
```{r}
is.data.frame(pbsc)

str(pbsc) # str() is a very useful function, you will use this function very often.

# dimension
dim(pbsc) # 55 rows and 5 columns
```


* Vectors
    + very basic data structure in R
    + two flavors: atomic vectors and lists.

```{r}
x = 1:10
x
is.vector(x)

# get a column from a data frame
pbsc$G
pbsc[, 1]
# a column from a data frame is an atomic vectors
is.vector(pbsc$G)
is.data.frame(pbsc$G)

# get a row from a data frame
pbsc[1, ]
# a row from a data frame is still a data frame.
is.vector(pbsc[1, ])
is.data.frame(pbsc[1, ])

# extract the column 'G' from a one-row data frame.
pbsc[1,]$G
```


* atomic vectors vs. lists
    + atomic vectors are homogeneous. They can only have one type of elements.
    + lists are homogeneous or heterogenous. They can have different types of elements. 
    
```{r}
my_vector = c(1, 3, 5) # c() function combine values into a vector.
my_vector

my_v2 = c('a', 'b', 'c')
my_v2

my_v3 = c(1, 2, 'b') # automatically convert numbers to strings.
my_v3

my_list = list(a=1:3, b=c('a', 'b', 'd'))
my_list
my_list$a

# any types of data can be an element of a list. let's add a data frame to our list.
my_l2 = list(a=1:3, b=c('a', 'b', 'd'), df=pbsc)
my_l2
my_l2$df
my_l2[1]
```


* matrix
    + two dimensions
    + **homogeneous**

```{r}
## we use the matrix() function to create a matrix
my_matrix = matrix(1:12, nrow=3, ncol=4) # a matrix with all numeric values
my_matrix

my_matrix_2 = matrix(letters[1:12], nrow=3, ncol=4) # a matrix with all character values
my_matrix_2

# convert a heterogeneous data frame to a homogeneous matrix
my_m3 = as.matrix(pbsc)
my_m3

# a column from a matrix is an vector
my_m3[, 3]
is.vector(my_m3[, 3])

# a row from a matrix is still an vector
my_m3[1, ]
is.vector(my_m3[1, ])
```

## Subsetting operators

There are three commonly used subsetting operators: `[[`, `$`, and `[`

* `$` is a shorthand for `[[`. It is useful when values are bound to character names. 
* `$` is invalid for atomic vectors.
* `[` is similar to `[[`. The difference is that `[[` can only return a **single element**, but `[[` can return **multiple elements**.

```{r}
# our data
pbsc
```

* ***A data frame is also a list.***
    + *the columns in the data frame are elements in the list.*


```{r}
is.data.frame(pbsc)
is.list(pbsc)
```


### subsetting atomic vectors

```{r}
# use a built-in dataset
letters 
my_v1 = letters[1:10] # get the first 10 elements in the vector letters
my_v1

# access single value
my_v1[1]
my_v1[[2]]

# access multiple values
my_v1[1:5] # first 5 elements
my_v1[c(1,5,9)] # elements 1, 5, and 9
my_v1[c(1,3,3,3,1:10)] # get some elements repeatedly! very useful!

my_v1[[1:5]] # this won't work

# access elements by names
names(my_v1) = LETTERS[1:10] # assign names to elements with the `names()` function
my_v1
my_v1['A'] # now we can access elements by element names.
my_v1[c('A','C','B')]

my_v1$A # this won't work! Because '$' does work for atomic vectors!
```

```{r}
# A use case for accessing elements by names.

my_sequence = sample(c('g', 'c', 't', 'a'), 20, replace = TRUE)
my_sequence

# how to convert lower case bases to upper case bases?
my_template = c('G','C','T','A')
names(my_template) = c('g', 'c', 't', 'a')
my_template

my_template['a'] 
my_template[c('g','g','a','c')] 

my_template[my_sequence] # yeah!
```


<hr>

### Exercises

* Create a few numeric or character vectors and try to extract some elements from the vectors.
* Assign names to vector elements and extract elements by names.

<hr>

### subsetting data frames

```{r}
# acess a single cell from data frame.
pbsc[1,2]
pbsc[[1,2]] # this will work since we are accessing single value.

# access multiple values
pbsc[1:5, ] # rows from 1 to 5
pbsc[-(1:5),] # remove the first five rows and keep the remaining
pbsc[, 1:2] # columns from 1 to 2
pbsc[1:5, 1:2] # subsetting by rows and columns

pbsc[[1:5, ]] # this won't work because we try to use '[[' to extract multiple elements!

# get one column from the data frame and remain the data structure.
pbsc[, 3]
is.vector(pbsc[, 3])

# what if we want to extract one column but keep the data frame structure.
pbsc[, 3, drop=FALSE]
is.vector(pbsc[, 3, drop=FALSE])
is.data.frame(pbsc[, 3, drop=FALSE])

# we can use the same position numbers repeatedly.
pbsc[, c(1:4, 4, 4)]
nrow(pbsc)
pbsc[c(1:2, 2, 2, 3:55), ]

# use $
pbsc$A
pbsc$X.Base
```


<hr>

### Exercise

* Get odd columns from the built-in datase **mtcars** (hint: you may need the `seq()` functin to create a integer sequence consisting of odd numbers. Use `?seq` to open the help documentation).

Your answer should look like this:

```
                    cyl  hp    wt vs gear
Mazda RX4             6 110 2.620  0    4
Mazda RX4 Wag         6 110 2.875  0    4
Datsun 710            4  93 2.320  1    4
Hornet 4 Drive        6 110 3.215  1    3
Hornet Sportabout     8 175 3.440  0    3
Valiant               6 105 3.460  1    3
Duster 360            8 245 3.570  0    3
Merc 240D             4  62 3.190  1    4
Merc 230              4  95 3.150  1    4
Merc 280              6 123 3.440  1    4
Merc 280C             6 123 3.440  1    4
Merc 450SE            8 180 4.070  0    3
Merc 450SL            8 180 3.730  0    3
Merc 450SLC           8 180 3.780  0    3
Cadillac Fleetwood    8 205 5.250  0    3
Lincoln Continental   8 215 5.424  0    3
Chrysler Imperial     8 230 5.345  0    3
Fiat 128              4  66 2.200  1    4
Honda Civic           4  52 1.615  1    4
Toyota Corolla        4  65 1.835  1    4
Toyota Corona         4  97 2.465  1    3
Dodge Challenger      8 150 3.520  0    3
AMC Javelin           8 150 3.435  0    3
Camaro Z28            8 245 3.840  0    3
Pontiac Firebird      8 175 3.845  0    3
Fiat X1-9             4  66 1.935  1    4
Porsche 914-2         4  91 2.140  0    5
Lotus Europa          4 113 1.513  1    5
Ford Pantera L        8 264 3.170  0    5
Ferrari Dino          6 175 2.770  0    5
Maserati Bora         8 335 3.570  0    5
Volvo 142E            4 109 2.780  1    4
```
<hr>

### subsetting list

```{r}
# create a list
my_l2 = list(a=1:3, b=c('a', 'b', 'd'), df=pbsc)
my_l2

# use $
my_l2$a
my_l2$df
```


```{r}
# use index to access multiple elements.
my_l2[1:3]
my_l2[c(1,1,1,1,2,2,3)]

my_l2[[1:3]] # this won't work!
```

**When `[` is applied to a list, it always returns a list. To access elements from within an element within a list, we need the `[[` operator.**

```{r}
my_l2[1]
is.list(my_l2[1])

# how to get the second value from element 'a' in list 'my_l2'
my_l2[1][2] # this won't work!!
my_l2[1][1][1][1][1][1][1][1][1][1][1][1][1][1][1][1]

my_l2[[1]]
is.list(my_l2[[1]])
is.vector(my_l2[[1]])

my_l2[[1]][2] # this will return the second value from element 'a' in list 'my_l2'
my_l2[[3]]
my_l2[[3]]$G
```


<hr>

### Exercise

* Extract the columns 'G' and 'C' from the data frame **pbsc** within the list **my_l2**.

<hr>

### subsetting matrix

```{r}
pbsc_matrix = as.matrix(pbsc)
pbsc_matrix

# access columns
pbsc_matrix[, 2]
pbsc_matrix[, 2:3]

# access rows
pbsc_matrix[1:2, ]

# '$' operator does not apply to matrix.
pbsc_matrix$G # this won't work!!

# get single column or rows and remain the data structure.
pbsc_matrix[1, ]
pbsc_matrix[1, ,drop=FALSE]

pbsc_matrix[ , 2]
pbsc_matrix[ , 2,drop=FALSE]
```


## Factors

When you import data from a file, R treats string (character) columns as factors by default.

```{r}
pbsc = read.csv('PBSC-comma.txt', header = TRUE)
pbsc$X.Base

# the stringsAsFactors argument
pbsc_str = read.csv('PBSC-comma.txt', header = TRUE, stringsAsFactors = FALSE)
pbsc_str$X.Base
```

Factors are used to store categorical data and are built on top of integer vectors.

You need two components to define a factor:

* A set of predefined categories (or **levels**)
* A vector that contain values only from those categories.

```{r}
# a set of categories: c('a', 'b', 'c')
# a vector: c('b', 'c', 'b', 'a', 'b')
factor_x1 = factor(levels = c('a', 'b', 'c'), x = c('b', 'c', 'b', 'a', 'b'))
factor_x1

# the vector does not have to contain all categories
factor_x2 = factor(levels = c('a', 'b', 'c'), x = c('b', 'b', 'b', 'b', 'b'))
factor_x2
# the vector need to contain values only from the predefined categories. Otherwise, it will generate NAs
factor_x2 = factor(levels = c('a', 'b', 'c'), x = c('b', 'b', 'b', 'b', 'b','d','e'))
factor_x2
```

**If you don't provide the category set, it will use the unique values from the vector as the category set**

```{r}
factor_x3 = factor(x = c('a', 'c', 'e', 'e', 'f', 'a'))
factor_x3
```


**How to understand that factors are built on top of integer vectors?**

```{r}
# the type of factors is 'integer'!
typeof(factor_x3)

factor_x1
as.integer(factor_x1)

factor_x3
as.integer(factor_x3)

# compare factors with string vectors
str_v1 = c('a', 'c', 'e', 'e', 'f', 'a')
as.integer(str_v1) 

str_v2 = c('1', '3', '4', '1') 
as.integer(str_v2)
```


**What happens if we combine two factors?**

```{r}
# combine two numeric vectors
num_x1 = c(1,2,3)
num_x2 = 5:10
c(num_x1, num_x2)

# combine two character vectors
str_x1 = c('a', 'b', 'c')
str_x2 = c('123', '45')
c(str_x1, str_x2)

# combine two factors
factor_x1 = factor(levels = c('high', 'low'), x = c('high', 'low','low', 'high', 'high', 'low'))
factor_x1
factor_x2 = factor(levels = c('medium', 'low'), x = c('medium', 'low','low', 'medium', 'low', 'low'))
factor_x2
c(factor_x1, factor_x2) # when combining two factors, factors are first converted to integer vectors and merged.

# how to get a merged factor as our result.
factors_list = list(factor_x1, factor_x2)
factors_list
unlist(factors_list)
```

**How are factor elements mapped to integers?**

* The factor elements are mapped to an integer **sequence that starts from 1**, and the mapping order is **determined by the order of factor levels**.


```{r}
factor_x1 = factor(levels = c('a', 'b', 'c'), x = c('b', 'c', 'b', 'a', 'b'))
factor_x1
as.integer(factor_x1)

factor_x4 = factor(levels = c('c', 'b', 'a'), x = c('b', 'c', 'b', 'a', 'b'))
factor_x4
as.integer(factor_x4)
as.integer(factor_x1)
```

Different level orders can result in different data analysis results due to the change in reference order. Next time when you do a linear regression analysis, try to reset the level orders for your categorical variables and see that happens.

**Set the reference level with `relevel()` function**

* the reference level matters when you dummy code a categorical variable for statistical analysis.
    
```{r}
## by default, the first level is the reference level.
factor_x4

factor_x4_ref = relevel(factor_x4, ref = 'b')
factor_x4_ref # now the level 'b' becomes the first and the others are moved down.
```



## Write data into files

* `write.table()`
* `write.csv()`
* `sink()`

```{r}
mtcars

write.table(x = mtcars, file = 'mtcars.txt', col.names = TRUE, row.names = TRUE)
write.csv(x = mtcars, file = 'mtcars.csv', col.names = TRUE) # generates a message
# let's open the help document for write.csv
?write.csv

write.csv(x = mtcars, file = 'mtcars.csv')

# recomend using write.table
# of course you can save the data into a csv file
write.table(x = mtcars, file = 'mtcars.csv', col.names = TRUE, row.names = TRUE)
write.table(x = mtcars, file = 'mtcars_2.csv', col.names = FALSE, row.names = FALSE)
```


`sink()` diverts R output to a connection.

```{r}
# open a connection by providing a file path
sink(file = 'my_linear_regression_analysis.txt')

# all outputs that usually go to your screen will be saved into the file.
lm(mpg~., data = mtcars) 
summary(lm(mpg~., data = mtcars))

# close the connection
sink()
```
